packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts

```diff
@@ -75,16 +101,172 @@ export class CheckpointService implements TreeDataProvider<BaseNode> {
+  public canAddCheckpointNote(): Boolean {
+    return this.getChildren().length < MAX_ALLOWED_CHECKPOINTS;
+  }
+
+  public async addCheckpointNode(
+    uriInput: string,
+    sourceFileInput: string,
+    typeRefInput: string,
+    lineInput: number,
+    makeCommandCall = true
+  ): Promise<CheckpointNode> {
+    // Before creating a new checkpoint node, check and see if one with the
+    // URI and line number already exists. If one already exists then return
+    // that one, otherwise create one, add it to the list and then return it
+    const cpTemp = this.returnCheckpointNodeIfAlreadyExists(
+      uriInput,
+      lineInput
+    );
+    if (cpTemp) {
+      return cpTemp;
+    }
+    const cpNode = new CheckpointNode(
+      uriInput,
+      sourceFileInput,
+      typeRefInput,
+      lineInput
+    );
+    // This is here for testing purposes. The command is being tested separately
+    if (makeCommandCall === true) {
+      await this.executeCreateApexExecutionOverlayActionCommand(cpNode);
+    }
     this.checkpoints.push(cpNode);
     this._onDidChangeTreeData.fire();
     return cpNode;
   }
+
+  // For checkpoints, there should only ever be one checkpoint for a given uri/line
+  // combination. ApexExecutionOverlayAction will not allow multiple actions for the
+  // same type/line.
+  private returnCheckpointNodeIfAlreadyExists(
+    uriInput: string,
+    lineInput: number
+  ): CheckpointNode | undefined {
+    for (const cp of this.checkpoints) {
+      if (cp.getCheckpointLineNumber() === lineInput) {
+        if (
+          cp.getCheckpointUri().toLocaleLowerCase() ===
+          uriInput.toLocaleLowerCase()
+        ) {
+          return cp;
+        }
+      }
+    }
+    // If we didn't find anything then just return undefined
+    return undefined;
+  }
+
+  public async deleteCheckpointNode(
+    uriInput: string,
+    lineInput: number,
+    makeCommandCall = true
+  ): Promise<boolean> {
+    const cpNode = this.returnCheckpointNodeIfAlreadyExists(
+      uriInput,
+      lineInput
+    );
+    if (cpNode) {
+      const index = this.checkpoints.indexOf(cpNode, 0);
+      if (index > -1) {
+        this.checkpoints.splice(index, 1);
+        this._onDidChangeTreeData.fire();
+        if (makeCommandCall === true) {
+          await this.executeRemoveApexExecutionOverlayActionCommand(cpNode);
+        }
+        return true;
+      }
+    }
+    return false;
+  }
+
+  // recreateIfDupe defaults to true, when the recurisve call is made
+  // this is set to false to prevent
+  private async executeCreateApexExecutionOverlayActionCommand(
+    theNode: CheckpointNode,
+    recreateIfDupe = true
+  ) {
+    // create the overlay action
+    const overlayActionCommand = new ApexExecutionOverlayActionCommand(
+      theNode.createJSonStringForOverlayAction()
+    );
+    let errorString = undefined;
+    let returnString = undefined;
+    await this.myRequestService.execute(overlayActionCommand).then(
```

Comment:
Using `await` and `then` together here is kind of weird. If the promise rejects, an error will be thrown and so you'll want to wrap this call in a try/catch block. If an error occurs with the request, catch it and do your error handling with the error string there

Reply:
James and I talked about this. It's staying. This looks to be the right call.